src/allmydata/web/root.py

```diff
@@ -29,30 +40,45 @@
     create_private_tree,
 )
 
-class URIHandler(RenderMixin, rend.Page):
-    # I live at /uri . There are several operations defined on /uri itself,
-    # mostly involved with creation of unlinked files and directories.
+class URIHandler(resource.Resource, object):
+    """
+    I live at /uri . There are several operations defined on /uri itself,
+    mostly involved with creation of unlinked files and directories.
+    """
 
     def __init__(self, client):
-        rend.Page.__init__(self, client)
+        super(URIHandler, self).__init__()
         self.client = client
 
-    def render_GET(self, ctx):
-        req = IRequest(ctx)
-        uri = get_arg(req, "uri", None)
-        if uri is None:
+    def render_GET(self, req):
+        """
+        Historically, accessing this via "GET /uri?uri=<capabilitiy>"
+        was/is a feature -- which simply redirects to the more-common
+        "GET /uri/<capability>" with any other query args
+        preserved. New code should use "/uri/<cap>"
+        """
+        uri_arg = get_arg(req, "uri", None)
+        if uri_arg is None:
             raise WebError("GET /uri requires uri=")
-        there = url.URL.fromContext(ctx)
-        there = there.clear("uri")
-        # I thought about escaping the childcap that we attach to the URL
-        # here, but it seems that nevow does that for us.
-        there = there.child(uri)
-        return there
-
-    def render_PUT(self, ctx):
-        req = IRequest(ctx)
-        # either "PUT /uri" to create an unlinked file, or
-        # "PUT /uri?t=mkdir" to create an unlinked directory
+        # so, using URL.from_text(req.uri) isn't going to work because
+        # it seems Nevow was creating absolute URLs including
+        # host/port whereas req.uri is absolute but lacks host/port
+        uri = URL.from_text(req.prePathURL().decode('utf8'))
+        # using ^ prePathURL() above because that includes the scheme
+        # / host / port but req.uri does not.
+        uri = uri.child(urllib.quote(uri_arg).decode('utf8'))
+        # add back all the query args that AREN'T ?uri=
+        for k, values in req.args.items():
+            if k != "uri":
+                for v in values:
+                    uri = uri.add(k.decode('utf8'), v.decode('utf8'))
+        return redirectTo(uri.to_text().encode('utf8'), req)
```

Comment:
This code now uses `hyperlink` .. perhaps not much shorter, but maybe "more likely to be correct"..?